src/core/services/interimElement/interimElementBase.js

```diff
@@ -0,0 +1,217 @@
+angular
+  .module('material.core')
+  .factory('$$interimElementBase', InterimElementBase);
+
+function InterimElementBase($mdUtil, $rootScope, $animate, $mdCompiler, $q, $$q, $mdTheming, $rootElement, $document,
+                            $timeout) {
+  return function(options) {
+    var API;
+    var element;
+    var showAction = $q.when(true);
+
+    prepareInterimElement();
+
+    return API = {
+      show: createAndShowElement,
+      remove: _runRemoveExecution,
+      options: options,
+      deferred: $q.defer()
+    };
+
+    function _runRemoveExecution(response, isCancelled, opts) {
+      // Abort and reject when the previous show failed.
+      if (!element) return $q.when(false);
+
+      options = angular.extend(options, opts || {});
+      options.cancelAutoHide && options.cancelAutoHide();
+      options.element.triggerHandler('$mdInterimElementRemove');
+
+      // When $destroy is set to true, then we are not waiting for the show transition to finish.
+      if (options.$destroy) {
+        return hideElement().then(function() {
+          (isCancelled && API.deferred.reject(response)) || API.deferred.resolve(response);
+        });
+      } else {
+
+        $q.when(showAction).finally(function() {
+          hideElement().then(function() {
+
+            (isCancelled && API.deferred.reject(response)) || API.deferred.resolve(response);
+
+          },  API.deferred.reject);
+        });
+
+        return API.deferred.promise;
+      }
+    }
+
+    function createAndShowElement() {
+      var defer = $q.defer();
+
+      compileElement(options).then(function(compiledData) {
+        element = linkElement(compiledData);
+
+        showAction = showElement(compiledData.controller).then(defer.resolve, rejectAll);
+      }, rejectAll);
+
+      function rejectAll(error) {
+        API.deferred.reject(error);
+        defer.reject(error);
+      }
+
+      return defer.promise;
+    }
+
+    function showElement(controller) {
+      var notifyShowing = options.onShowing || angular.noop;
+      var notifyComplete = options.onComplete || angular.noop;
+
+      // Notify the `onShowing` function before the show process starts.
+      notifyShowing(options.scope, element, options, controller);
+
+      var defer = $q.defer();
+
+      try {
+        $q.when(options.onShow(options.scope, element, options, controller))
+          .then(function() {
+
+            notifyComplete(options.scope, element, options);
+            startAutoHide();
+
+            defer.resolve(element);
+
+          }, defer.reject)
+
+      } catch (err) {
+        defer.reject(err.message);
+      }
+
+      return defer.promise;
+    }
+
+    function hideElement() {
+      var defer = $$q.defer();
```

Comment:
@crisbeto Great question, had the same question today and asked Thomas about that.

Quote from Thomas:
> $$q does NOT depend upon the $digest() cycle nor does it trigger $digest()

See `$$q` in the source. https://github.com/angular/angular.js/blob/master/src/ng/q.js#L228